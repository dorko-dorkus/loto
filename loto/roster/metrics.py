"""Utility models and helpers for ranking field operations.

This module defines small Pydantic models used by the roster portion of the
project as well as a couple of metric helpers.  The goal is to keep the code
lightweight and dependency free while still exercising a few common statistical
patterns: exponentially weighted moving averages, Bayesian shrinkage and
threshold based ranking.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal, Sequence, Tuple

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Pydantic models
# ---------------------------------------------------------------------------


class KpiEvent(BaseModel):
    """Represents a single KPI measurement.

    Attributes
    ----------
    timestamp:
        When the event occurred.
    incidents:
        Number of incidents observed.
    total:
        Total population for the KPI (denominator for a rate).
    """

    timestamp: datetime = Field(..., description="When the KPI was measured")
    incidents: int = Field(0, ge=0, description="Number of incidents")
    total: int = Field(1, ge=1, description="Population for the KPI")

    class Config:
        extra = "forbid"


class HatScore(BaseModel):
    """Score associated with a hat ranking system."""

    value: float = Field(..., ge=0.0, le=1.0, description="Score between 0 and 1")
    incidents: int = Field(0, ge=0)
    total: int = Field(0, ge=0)

    class Config:
        extra = "forbid"


class HatRank(BaseModel):
    """Represents the discrete band a score falls into."""

    band: Literal["green", "amber", "red"]
    score: float = Field(..., ge=0.0, le=1.0)

    class Config:
        extra = "forbid"


# ---------------------------------------------------------------------------
# Metric helpers
# ---------------------------------------------------------------------------


def ewma(previous: float, observation: float, dt: float, half_life: float) -> float:
    """Update an exponentially weighted moving average.

    Parameters
    ----------
    previous:
        The previous average value.
    observation:
        The new observation to incorporate.
    dt:
        Time since the last observation in the same units as ``half_life``.
    half_life:
        The period over which the influence of ``previous`` decays by half.

    Returns
    -------
    float
        The updated moving average.
    """

    if half_life <= 0:
        raise ValueError("half_life must be positive")

    # Convert the half-life to an exponential smoothing factor.
    alpha = 1 - 0.5 ** (dt / half_life)
    return (1 - alpha) * previous + alpha * observation


def shrink(
    incidents: int,
    total: int,
    *,
    prior: float = 0.5,
    weight: float = 1.0,
    cap: int | None = None,
) -> float:
    """Bayesian shrinkage of an incident rate.

    A simple Beta prior centred on ``prior`` is blended with the observed
    incident rate.  ``weight`` controls the strength of the prior and ``cap``
    places an upper bound on the number of incidents considered.
    """

    if total < 0:
        raise ValueError("total must be non-negative")
    if incidents < 0:
        raise ValueError("incidents must be non-negative")
    if cap is not None:
        incidents = min(incidents, cap)

    numerator = incidents + prior * weight
    denominator = total + weight
    return numerator / denominator if denominator else prior


def rank_bands(score: float, bands: Sequence[Tuple[float, str]]) -> str:
    """Return the rank band name for ``score``.

    ``bands`` should be an iterable of ``(threshold, name)`` pairs ordered from
    highest to lowest threshold.  The first band whose threshold is less than or
    equal to ``score`` is returned.
    """

    for threshold, name in bands:
        if score >= threshold:
            return name
    # Fall back to the lowest band name if nothing matches
    return bands[-1][1]
