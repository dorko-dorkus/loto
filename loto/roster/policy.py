from __future__ import annotations

from pathlib import Path
from typing import Any

import yaml

_DEFAULT_PATH = Path(__file__).resolve().parents[2] / "config" / "hats_policy.yaml"


def load_policy(path: str | Path = _DEFAULT_PATH) -> dict[str, Any]:
    """Return policy configuration from *path*.

    The file is expected to contain YAML with keys defining ranking and chooser
    behaviour.  The returned mapping is normalised to the keys used by the
    ranking and scheduling helpers.
    """

    data = yaml.safe_load(Path(path).read_text()) or {}
    policy: dict[str, Any] = {
        "weights": data.get("weights"),
        "half_life": data.get("half_life"),
        "pseudo_count": data.get("pseudo_count"),
        "incident_cap": data.get("incident_cap"),
        "rotation_limit": data.get("rotation_window"),
        "utilization_cap": data.get("daily_utilization_cap"),
    }
    return policy
