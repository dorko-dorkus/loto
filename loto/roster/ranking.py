"""Ranking algorithms for roster metrics.

This module exposes :func:`update_ranking` which consumes a *ledger* – a
mapping of entity identifiers to an ordered sequence of metric tuples – and
returns a snapshot containing the rank and coefficient for each entity.

The calculation demonstrates a few statistical techniques:

* **Composite** – multiple metrics for a single observation are combined into
  a single score by averaging them.
* **EWMA** – the sequence of composite scores is reduced to a single value via
  an exponential weighted moving average.
* **Shrinkage** – each EWMA value is pulled slightly towards the global mean to
  temper extreme outliers.
* **Bands** – the final coefficient is categorised into a band (``S`` through
  ``C``) based on simple thresholds.
* **Caps** – coefficients are clipped into a fixed range to keep them
  well‑behaved.

The function is intentionally small and self contained so it can serve as a
teaching example for statistical style ranking pipelines.
"""

from __future__ import annotations

from typing import Dict, Mapping, Sequence

_ALPHA = 0.5
_SHRINKAGE = 0.1
_THRESHOLDS = (0.75, 0.5, 0.25)  # S, A, B, else C
_CAPS = (0.0, 1.0)


def _composite(observation: Sequence[float]) -> float:
    """Combine multiple metrics into a single score.

    A simple arithmetic mean is used which is sufficient for the tests and keeps
    the function intentionally straightforward.
    """

    if not observation:
        return 0.0
    return float(sum(observation) / len(observation))


def _ewma(values: Sequence[float], alpha: float = _ALPHA) -> float:
    """Compute the exponential weighted moving average for *values*."""

    iterator = iter(values)
    try:
        acc = float(next(iterator))
    except StopIteration:
        return 0.0
    for value in iterator:
        acc = alpha * float(value) + (1.0 - alpha) * acc
    return acc


def _apply_shrinkage(value: float, mean: float, factor: float = _SHRINKAGE) -> float:
    """Pull *value* towards *mean* by ``factor``."""

    return (1.0 - factor) * value + factor * mean


def _band(value: float) -> str:
    """Return the categorical band for *value*."""

    t_s, t_a, t_b = _THRESHOLDS
    if value >= t_s:
        return "S"
    if value >= t_a:
        return "A"
    if value >= t_b:
        return "B"
    return "C"


def update_ranking(
    ledger: Mapping[str, Sequence[Sequence[float]]],
) -> Dict[str, Dict[str, object]]:
    """Return a ranking snapshot for the given *ledger*.

    Parameters
    ----------
    ledger:
        Mapping of entity identifiers to an ordered sequence of observations.
        Each observation may contain one or more metrics.  Each metric tuple is
        collapsed into a single composite value prior to further calculations.

    Returns
    -------
    dict
        A mapping of entity identifier to a dictionary containing ``rank``,
        ``coefficient`` and ``band`` entries.
    """

    # Step 1 – build composite score history for each entity.
    histories: Dict[str, list[float]] = {
        name: [_composite(obs) for obs in observations]
        for name, observations in ledger.items()
    }

    # Step 2 – compute EWMA for each entity.
    ewmas: Dict[str, float] = {name: _ewma(vals) for name, vals in histories.items()}

    # Step 3 – shrink towards global mean to reduce variance.
    global_mean = sum(ewmas.values()) / len(ewmas) if ewmas else 0.0
    coefficients: Dict[str, float] = {
        name: _apply_shrinkage(value, global_mean) for name, value in ewmas.items()
    }

    # Step 4 – apply caps.
    cap_min, cap_max = _CAPS
    coefficients = {
        name: min(max(coeff, cap_min), cap_max) for name, coeff in coefficients.items()
    }

    # Step 5 – assign bands and ranks.
    ranked_names = sorted(coefficients, key=lambda n: coefficients[n], reverse=True)
    snapshot: Dict[str, Dict[str, object]] = {}
    for position, name in enumerate(ranked_names, start=1):
        coeff = round(coefficients[name], 4)
        snapshot[name] = {
            "rank": position,
            "coefficient": coeff,
            "band": _band(coeff),
        }

    return snapshot
