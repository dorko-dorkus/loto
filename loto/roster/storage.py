from __future__ import annotations

import hashlib
import json
from pathlib import Path
from typing import Any, Dict, Iterable, List, Mapping, cast


def _entry_hash(wo_id: str, hat_id: str) -> str:
    """Return a deterministic hash for the given work order and hat IDs."""
    data = f"{wo_id}:{hat_id}".encode("utf-8")
    return hashlib.sha256(data).hexdigest()


def read_ledger(path: Path) -> List[Dict[str, Any]]:
    """Return all ledger entries stored at ``path``.

    The ledger is stored as a JSON Lines (JSONL) file where each line is a
    JSON object representing a single entry. Missing files yield an empty
    list.
    """
    if not path.exists():
        return []
    entries: List[Dict[str, Any]] = []
    with path.open("r", encoding="utf-8") as fh:
        for line in fh:
            line = line.strip()
            if line:
                entries.append(json.loads(line))
    return entries


def append_ledger(path: Path, entry: Mapping[str, Any]) -> Dict[str, Any]:
    """Append ``entry`` to the ledger at ``path``.

    The ledger is append-only. If an entry with the same ``wo_id`` and
    ``hat_id`` already exists the write is refused and :class:`ValueError`
    is raised.
    """
    wo_id = str(entry["wo_id"])
    hat_id = str(entry["hat_id"])
    key = _entry_hash(wo_id, hat_id)
    for existing in read_ledger(path):
        existing_key = _entry_hash(str(existing["wo_id"]), str(existing["hat_id"]))
        if existing_key == key:
            raise ValueError("duplicate ledger entry")
    path.parent.mkdir(parents=True, exist_ok=True)
    to_write: Dict[str, Any] = dict(entry)
    to_write["_id"] = key
    with path.open("a", encoding="utf-8") as fh:
        fh.write(json.dumps(to_write) + "\n")
    return to_write


def compute_snapshot(entries: Iterable[Mapping[str, Any]]) -> Dict[str, Dict[str, Any]]:
    """Compute a snapshot from ledger ``entries``.

    The snapshot is a dictionary keyed by the hash derived from ``wo_id``
    and ``hat_id``. The latest occurrence of a key wins.
    """
    snapshot: Dict[str, Dict[str, Any]] = {}
    for entry in entries:
        key = _entry_hash(str(entry["wo_id"]), str(entry["hat_id"]))
        snapshot[key] = dict(entry)
    return snapshot


def write_snapshot(path: Path, snapshot: Dict[str, Dict[str, Any]]) -> None:
    """Write ``snapshot`` to ``path`` as JSON."""
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8") as fh:
        json.dump(snapshot, fh)


def read_snapshot(path: Path) -> Dict[str, Dict[str, Any]]:
    """Read snapshot data from ``path``.

    Missing files or empty contents yield an empty dictionary.
    """
    if not path.exists():
        return {}
    content = path.read_text(encoding="utf-8").strip()
    if not content:
        return {}
    return cast(Dict[str, Dict[str, Any]], json.loads(content))
