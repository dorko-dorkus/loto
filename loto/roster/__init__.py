"""Roster ranking utilities."""

from .ranking import update_ranking

__all__ = ["update_ranking"]
