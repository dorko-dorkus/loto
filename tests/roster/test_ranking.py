from __future__ import annotations

import json
from pathlib import Path

import pytest

from loto.roster.ranking import update_ranking


@pytest.fixture
def golden(request):
    path = Path(request.node.fspath).with_suffix(".golden.json")
    expected = json.loads(path.read_text())

    def check(data: object) -> None:
        assert data == expected

    return check


def test_update_ranking(golden):
    ledger = {
        "alice": [(0.7, 0.8), (0.8, 0.9), (0.9, 1.0)],
        "bob": [(0.5, 0.4), (0.4, 0.5), (0.6, 0.5)],
        "carol": [(0.2, 0.1), (0.3, 0.2), (0.4, 0.3)],
    }

    snapshot = update_ranking(ledger)
    golden(snapshot)
