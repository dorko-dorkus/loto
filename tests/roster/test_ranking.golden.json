{
  "alice": {
    "band": "S",
    "coefficient": 0.8425,
    "rank": 1
  },
  "bob": {
    "band": "A",
    "coefficient": 0.505,
    "rank": 2
  },
  "carol": {
    "band": "B",
    "coefficient": 0.3025,
    "rank": 3
  }
}
